### PR TITLE
Remove complex cache, add less complex cache for event types

### PR DESCRIPTION
**Pull request by antban**
ARUHA-2695

Avoid running huge database requests

### CODE REVIEW

File: core-services/src/main/java/org/zalando/nakadi/cache/EventTypeCache.java [hunk 1]
```diff
@@ -1,101 +1,195 @@
...
-
+            final TimelineSync timelineSync,
+            final EventValidatorBuilder eventValidatorBuilder,
+            @Value("${nakadi.event-cache.periodic-update-seconds:120}") final long periodicUpdatesIntervalSeconds,
+            @Value("${nakadi.event-cache.change-ttl:600}") final long zkChangesTTLSeconds) {
+        this.changesRegistry = changesRegistry;
+        this.eventTypeDbRepository = eventTypeDbRepository;
         this.timelineRepository = timelineRepository;
+        this.valueCache = CacheBuilder.newBuilder()
+                .expireAfterWrite(Duration.ofHours(2))
```
Comment (rcillo) [inline core-services/src/main/java/org/zalando/nakadi/cache/EventTypeCache.java]: I expected `expireAfterWrite` to be set to `periodicUpdatesIntervalSeconds`. This would guarantee that data is forcefully fetched from database even in the case an "update broadcast" notification is lost via zookeeper.

Then you could possibly eliminate the need for the scheduled `periodicCheck`.

Sorry if I didn't understand something. Maybe it would be worth documenting the behavior of the cache in plain english as a class documentation.

File: core-services/src/main/java/org/zalando/nakadi/cache/ChangeSet.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+                final boolean alreadyExists = currentEtChanges.stream().anyMatch(c -> c.getId().equals(ch.getId()));
+                if (!alreadyExists) {
+                    eventTypes.add(ch.getEventTypeName());
+                }
+            }
+        }
+        return eventTypes;
+    }
+
+    private Date getActualChangeDate(final Change change) {
```
Comment (rcillo) [inline core-services/src/main/java/org/zalando/nakadi/cache/ChangeSet.java]: What do you mean by `actual`? What could it be if not `actual`?

**Comment by rcillo** [comment]
@antban I was expecting something a little different. I thought the new cache behaviour would be as follow:

- eviction based on zookeeper watch (to possibly speedup eviction, not having to wait for 2 minutes for propagation of changes)
- forceful eviction time of 2 minutes. This is because watches don't provide strong guarantees that all changes will be observed and this way we at least have guaranteed eventual consistency of 2 minutes

So in the code I expected to see some simple watcher on top of some znode to force eviction and the code to reset this watcher. I also expected some simple eviction time. But the code has a lot more than I expected. Maybe I need some in person explanation on why you considered this way to be better.

**Comment by antban** [comment]
@rcillo I will write description of the way how it works, and than we could discuss it in person

**Comment by antban** [comment]
@rcillo Added some documentation to event type cache https://github.com/zalando/nakadi/pull/1154/commits/4258e4b4244948e3c65b218571d0fa6a43498e21

**Comment by antban** [comment]
:+1:

**Comment by antban** [comment]
@rcillo Refactored to support TTL, therefore removed part of the code that was handling deletions: 
https://github.com/zalando/nakadi/pull/1154/commits/48007e4405e250fdebb7191de9b79be93a2fd4ce

**Comment by antban** [comment]
:-1: Please do not merge it cause we have some dependencies to be updated beforehands

**Comment by antban** [comment]
:+1:

**Comment by rcillo** [comment]
👍

File: core-services/src/main/java/org/zalando/nakadi/cache/ChangesRegistry.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    @PostConstruct
+    public void postConstruct() {
+        try {
+            zk.get().create().creatingParentsIfNeeded().forPath(ET_CACHE_PATH);
+        } catch (KeeperException.NodeExistsException ex) {
+            // Its all fine
+        } catch (Exception ex) {
+            throw new NakadiRuntimeException(ex);
```
[the remark this answers is not in the document]
Comment (antban) [reply core-services/src/main/java/org/zalando/nakadi/cache/ChangesRegistry.java]: @rcillo Regarding this point - I think it makes sense not to retry, but just restart, as it may lead to  further problems during initialization.
I mean right now our code uses this kind of approach everywhere - we are retrying only in case if already started. During initialization - not. I thought to leave it like that to avoid confusion with other places

[the remark this answers is not in the document]
Comment (rcillo) [reply core-services/src/main/java/org/zalando/nakadi/cache/ChangesRegistry.java]: Agreed.
